### DISCUSSION TITLE
Closed Source Vs Open Source Software: Focusing on Security

### DISCUSSION THREAD

**Discussion by Wny-Duong** [albertkun/21S-ASIAAM-191A#113] [Past Thinking Caps]
There's always been a history of debate regarding the security of software that originates from being open source vs closed source. 

A very common, and even reasonable claim, is that open source software tends to provide potential bad agents with the information they need to attack vulnerabilities in software. Things like digital signatures from patches, access to the code directly, and frequency of cyber attacks on open source tools tends to create a persuasive case that closed source software is more secure than open source [1]. A counter-response to those things would be arguments that closed-source software is also similarly vulnerable due to the ability to reverse-engineer such software to acquire the source code through automated attacking tools. Adding more in favor of open-source, there is also the matter of security being given greater scrutiny (given a reasonably large development community) as a result of the number of developers working on open source software. 

One of the biggest weaknesses/strengths of closed source comes from the fact that it is software often developed upon by a small team where the security benefit comes from having as few eyes as possible on the codes, protecting it from potential malicious agents. This is weirdly enough, both a good and a bad thing [2]. It's a good thing to people who look on that development team to be a reliable and competent source of security, which means having a lot of faith in the software. But, having trust in software isn't enough to give it some tangible proof of safety. This code obscurity becomes a bad thing when developers believe that it's more necessary to have more eyes on code, which means that the security comes from a large community being able to identify issues and vulnerabilities in it. 

This is idea is somewhat reinforced by earlier research done on Github development communities vulnerability of software [3]. The findings brought up in that research suggest that it's often the latter case, that closed source software's obscurity is an issue. A majority of security vulnerabilities is found to have originated from development issues and mistakes, as opposed to some large hole in security that was actually exploited. If we consider cases like this to be representative of standard open and closed source software development, it'd mean that open source would be a better methodology to develop software because it allows the majority case of vulnerabilities to be addressed. If the majority of vulnerabilities is caused by human error and not malicious exploitation of code, then it fits that having more people look and bring oversight to code would help flesh those vulnerabilities out.

Source:
[1] https://www.technologyreview.com/2010/06/07/262916/open-source-could-mean-an-open-door-for-hackers/
[2] https://www.researchgate.net/publication/228954879_OPEN_SOURCE_VS_CLOSED_SOURSE
[3] https://octoverse.github.com/

**Comment by jsxu315** [comment]
You bring up some really good nuances of security in open- vs closed-source! I found it really interesting that closed-source software's biggest vulnerabilities stem from the human component and not purely technical aspects. 

This also made me think about how the image that companies put out might lead people to put (perhaps misplaced) trust in companies to protect against vulnerabilities. Thinking about users' experiences, another potential advantage of open-source over closed-source here is that if a vulnerability were ever exploited, open-source devs might give the user community more help and clarity than a company that wants to protect its image.

**Comment by rh5140** [comment]
What a thorough response! I knew about the security concerns you described with open source software, but I didn't realize the ones you bring up with closed source. The flip side of open source software being more easily accessible to bad actors is that it's more likely that such a security issue would be identified quickly.

> This code obscurity becomes a bad thing when developers believe that it's more necessary to have more eyes on code, which means that the security comes from a large community being able to identify issues and vulnerabilities in it.

I like this sentence a lot since I feel like it concisely captures the main point you were making about security issues with closed source!

Your point about people having faith that a software will be secure is interesting too. From that, just from a public usage perspective, closed source is most viable for already-well-established companies (read: probably monopolies) while smaller development teams might be able to foster more trust in their target audience by making their software open source.
